packages/ember-glimmer/tests/integration/helpers/custom-helper-test.js

```diff
@@ -22,4 +26,431 @@ moduleFor('Helpers test: custom helpers', class extends RenderingTest {
+      destroy() {
+        destroyCalled++;
+        this._super(...arguments);
+      },
+      compute() {
+        return 'must define a compute';
+      }
+    });
+
+    this.render(`{{some-helper}}`);
+
+    runDestroy(this.component);
+
+    assert.strictEqual(destroyCalled, 1, 'destroy called once');
+  }
+
+  ['@test helper used in subexpression can recompute']() {
+    let helper;
+    let phrase = 'overcomes by';
+
+    this.registerHelper('dynamic-segment', {
+      init() {
+        this._super(...arguments);
+        helper = this;
+      },
+      compute() {
+        return phrase;
+      }
+    });
+
+    this.registerHelper('join-words', {
+      compute(params) {
+        return params.join(' ');
+      }
+    });
+
+    this.render(
+      `{{join-words "Who"
+                   (dynamic-segment)
+                   "force"
+                   (join-words (join-words "hath overcome but" "half"))
+                   (join-words "his" (join-words "foe"))}}`);
+
+    this.assertText('Who overcomes by force hath overcome but half his foe');
+
+    this.runTask(() => this.rerender());
+
+    this.assertText('Who overcomes by force hath overcome but half his foe');
+
+    phrase = 'believes his';
+
+    this.runTask(() => helper.recompute());
+
+    this.assertText('Who believes his force hath overcome but half his foe');
+
+    phrase = 'overcomes by';
+
+    this.runTask(() => helper.recompute());
+
+    this.assertText('Who overcomes by force hath overcome but half his foe');
+  }
+
+  ['@test helper used in subexpression can recompute component']() {
+    let helper;
+    let phrase = 'overcomes by';
+
+    this.registerHelper('dynamic-segment', {
+      init() {
+        this._super(...arguments);
+        helper = this;
+      },
+      compute() {
+        return phrase;
+      }
+    });
+
+    this.registerHelper('join-words', {
+      compute(params) {
+        return params.join(' ');
+      }
+    });
+
+    this.registerComponent('some-component', {
+      template: '{{first}} {{second}} {{third}} {{fourth}} {{fifth}}'
+    });
+
+    this.render(
+      `{{some-component first="Who"
+                   second=(dynamic-segment)
+                   third="force"
+                   fourth=(join-words (join-words "hath overcome but" "half"))
+                   fifth=(join-words "his" (join-words "foe"))}}`);
+
+    this.assertText('Who overcomes by force hath overcome but half his foe');
+
+    this.runTask(() => this.rerender());
+
+    this.assertText('Who overcomes by force hath overcome but half his foe');
+
+    phrase = 'believes his';
+
+    this.runTask(() => helper.recompute());
+
+    this.assertText('Who believes his force hath overcome but half his foe');
+
+    phrase = 'overcomes by';
+
+    this.runTask(() => helper.recompute());
+
+    this.assertText('Who overcomes by force hath overcome but half his foe');
+  }
+
+  ['@htmlbars helper used in subexpression is destroyed']() {
```

Comment:
This is marked as `@htmlbars` because it does not call `destroy` in Glimmer2, but it should.